### DISCUSSION TITLE
Very slow performance

### DISCUSSION THREAD

**Discussion by yondr-agency-15** [statamic/cms#5213] [Troubleshooting]
Hello there,

We are having some issues with loading times on our statamic website.
I've noticed that in general statamic is very slow when not using it's caching system.

### Production:

we have enabled the following settings (env):
````
DEBUGBAR_ENABLED=false`
STATAMIC_STATIC_CACHING_STRATEGY=full
STATAMIC_STACHE_WATCHER=false
````

### The issue

If the cache was cleared and we try to access some pages we'll get the following message:
Request Timeout
`This request takes too long to process, it is timed out by the server. If it should not be timed out, please contact administrator of this web site to increase 'Connection Timeout`.

The main issue is that when the page has not been cached yet, it take very long to create cached/static version of it.

### Page reload (non-cached)
On local development when I turn off my `STATAMIC_STATIC_CACHING_STRATEGY`. 
A page reload takes at least 5 seconds up to even 30+ seconds. (even for a minimal contact page)
minimal page : ![image](https://user-images.githubusercontent.com/93581121/152823281-8fbf6bca-660a-43af-892d-da24e275d0e4.png)
bigger page:  ![image](https://user-images.githubusercontent.com/93581121/152825025-d6436d52-b463-4c1f-a3b4-6eada03d8e47.png)

### Control panel
On our staging sometimes the "control panel" is also verry slow and we get the following message (not sure if this is associated with pageload on my website): `Too many requests to statamic.com. Try again in 18 seconds.`

**Comment by duncanmcclean** [comment]
How many entries do you have? If it's > 3k then I'd look into static caching.

**Reply by yondr-agency-15** [reply]
All less than 40
![image](https://user-images.githubusercontent.com/93581121/152832368-333c7d0b-7e52-4844-a3ae-da71febe5d17.png)

**Comment by jasonvarga** [comment]
We'll need more information about your setup. What does your template look like?

**Reply by yondr-agency-15** [reply]
Hey Jason,

it's an inherited project so I'm still figuring some thing out but the current setup is:

### Setup
Statamic **V3.2** using the regular **antler** template engine.  (Without pro license atm)
There aren't a lot of **plugins** installed, just the: **duplicator, seo pro & cache controller**.


### Example:
Here is an example of our contact page:

<details>
  <summary>contact.antlers.html</summary>
  
```
 {{ section:body_classes }}is-bg-grey{{ /section:body_classes }}

<section class="xs:min-h-[50vh]">
    <div class="container">
        
        <div class="grid" data-visible-repeat="up">
            <div class="grid-col w-full xs:w-1/2" data-visible-repeat-text>
                {{ general:address | nl2br }} <br>
                <a href="tel:{{ general:phonenumber }}">{{ general:phonenumber }}</a>
            </div>

            <div class="grid-col w-full xs:w-1/2 mt-md sm:mt-0" data-visible-repeat-text>
                {{ general:business_info | nl2br }} <br>
            </div>

            {{ general:emails }}
                <div class="grid-col w-full xs:w-1/2 mt-md sm:mt-xl" data-visible-repeat-text>
                    <p class="text-underline">{{ label }}</p>
                    <a class="h2 | link-arrow" href="mailto:{{ email }}">
                        <span class="link-arrow__label">
                            {{ email }}
                        </span>
                        <span class="link-arrow__arrow">↘</span>
                    </a>
                </div>
            {{ /general:emails }}

            <div class="grid-col w-full xs:w-1/2 mt-md sm:mt-xl" data-visible-repeat-text>
                <p class="text-underline">{{ vacancies_title }}</p>
                {{ partial:components/button :button="vacancies_button" arrow="true" }}
            </div>
        </div>

    </div>
```

</details>

**Edit:**  I,ve just tested to load an empty page and this also take 3seconds to load. 


### Extra:
There is also usage of tailwindCSS and Alpinejs. 
For local development I use a wamp server and run the `mix watch --hot` command.

**Reply by jasonvarga** [reply]
If you create a completely fresh site using `statamic new mysite`... is that slow too?

**Reply by yondr-agency-15** [reply]
Fresh template does not seem to be slow.

**Comment by jackmcdade** [comment]
Are you using the Nav tag, by chance?

**Reply by yondr-agency-15** [reply]
Like the html `<nav></nav>` tag? yes I am.
And yes also the antlers nav tag  `{{ nav:main as="links" }} {{ /nav:main }}`

**Reply by jackmcdade** [reply]
Sorry yeah the {{ nav }} tag. Can you try to take it out as an experiment and see how much faster your site is?

**Reply by yondr-agency-15** [reply]
No, same result without the nav tag :/

**Comment by yondr-agency-15** [comment]
I've just discovered that using Laragon instead of Wamp makes a huge difference. My Largest page went from down to a ~ 2s loading time.
So it probably has to do something with some server settings, any ideas? Suggestions? Because it's still an issue at our staging website.

**EDIT: ** after clearing the cache page's, heavy pages still take 30+ seconds, but a tiny contact page is now ready under 1 second.

**Reply by duncanmcclean** [reply]
Yeah, your server environment can make a difference to loading speeds, as you can tell. What's your staging server setup?

**Reply by yondr-agency-15** [reply]
We're using cPanel. It seems to be running on a zend-server v3.4.0 using php v7.4.16.

**Comment by eliasu** [comment]
Did you solve this or improve the performance? I also wondered if i'm overusing the Antler conditions or something …

**Reply by ExpDev07** [reply]
Doing the following brought my loading time down to < 300ms:

- Disable file watching.
- Enable half static caching (or use cache tags around navs and replicators). 
- Disable debugbar.

**Comment by lazdins** [comment]
I was given to maintain a website. Running V3.3.68 Pro, the performance is very slow. According to dev bar very request is more than 3.5 s to the page load. Other partials where nav is used makes ir even slower. I tried to cache everything but that is not the solution for our case as there are dynamic is active and is parent. What to do? This is the _sidebar.antlers.html https://pastebin.com/v8JD9j5m The file watcher is disabled and had full caching setup in .env if that changes something. The total count of contents in collections is 4k+

**Reply by duncanmcclean** [reply]
Have you tried out the [Eloquent driver](https://github.com/statamic/eloquent-driver) to see if that improves performance?

**Reply by lazdins** [reply]
No. I was hoping that previous devs done something terribly wrong in the antlers but it seems that Statamic is just slow for such queries

**Comment by jackmcdade** [comment]
Did you add an index for the fields you’re filtering on?

**Reply by lazdins** [reply]
Where to check this? I’m new to the Statamic, I have been assigned to maintain an existing website but the performance issues are killing the production server each day.

**Reply by jackmcdade** [reply]
Check out this section of the docs: https://statamic.dev/stache#indexes

**Reply by lazdins** [reply]
Hi, I tried to add indexes In /statamic/config/statamic/stache.php in place of original     
```
'indexes' => [
        //
    ],
```

tried to add

```
'indexes' => [
    'navigation' => [
        'title' => \Statamic\Stache\Indexes\Value::class,
        'url' => \Statamic\Stache\Indexes\Value::class,
    ],
],
```

When trying to php please stache:clear or stache:warm I am getting:
```
   Error 

  Class name must be a valid object or a string

  at vendor/statamic/cms/src/Stache/Stores/Store.php:59
     55▕         }
     56▕ 
     57▕         $class = $this->indexes()->get($name, $this->valueIndex);
     58▕ 
  ➜  59▕         $index = new $class($this, $name);
     60▕ 
     61▕         $cached->put($key, $index);
     62▕ 
     63▕         return $index;

      +1 vendor frames 
  2   [internal]:0
      Statamic\Stache\Stores\Store::Statamic\Stache\Stores\{closure}()

      +22 vendor frames 
  25  artisan:37
      Illuminate\Foundation\Console\Kernel::handle()
```

What could be wrong? Statamic Pro 3.3.68

**Comment by jackmcdade** [comment]
I think it's just this:

```
'indexes' => [
    'navigation' => [
        'title',
        'url'
    ],
],
```

**Reply by lazdins** [reply]
Thanks but got the same error:
  ```
 Error 

  Class name must be a valid object or a string

  at vendor/statamic/cms/src/Stache/Stores/Store.php:59
     55▕         }
     56▕ 
     57▕         $class = $this->indexes()->get($name, $this->valueIndex);
     58▕ 
  ➜  59▕         $index = new $class($this, $name);
     60▕ 
     61▕         $cached->put($key, $index);
     62▕ 
     63▕         return $index;

      +1 vendor frames 
  2   [internal]:0
      Statamic\Stache\Stores\Store::Statamic\Stache\Stores\{closure}()

      +22 vendor frames 
  25  please:37
      Illuminate\Foundation\Console\Kernel::handle()
```

**Comment by edalzell** [comment]
Let's see your full `stache.php` please.

**Reply by lazdins** [reply]
Hi, this is the current version of full config/statamic/stache.php - https://pastebin.com/ciUCQ7PR
